{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "from IPython.display import display\n",
    "\n",
    "import duckdb\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import logging\n",
    "from Bio import SeqIO\n",
    "from datetime import datetime\n",
    "import time\n",
    "\n",
    "from typing import List, Optional, Union, Dict, Set\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "logger = logging.getLogger(__name__)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build initial database\n",
    "\n",
    "We're using our `database.builder` module to build the database.\n",
    "\n",
    "TODO: Add logic to only insert sequences that don't already exist in the database.\n",
    "\n",
    "TODO: I think there's some duplication/differences happening in the context manager that seem dumb, both of the \"with\" statements should be able to be combined."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:planter.database.builder:Applying migration: 000_schema_version.sql\n",
      "INFO:planter.database.builder:Applying migration: 001_initial_schema.sql\n",
      "INFO:planter.database.builder:Applying migration: 002_add_indexes.sql\n",
      "INFO:planter.database.builder:Fetching metadata for ERR9123871\n",
      "INFO:planter.database.utils.sra:Fetching metadata for ERR9123871\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:planter.database.builder:Loading sequences from /mnt/data2/planter_outputs/ERR9123871/transdecoder/ERR9123871.pep\n",
      "INFO:planter.database.builder:Loaded 30281 new sequences for ERR9123871\n",
      "INFO:planter.database.builder:Loading annotations from /mnt/data2/planter_outputs/ERR9123871/eggnog/ERR9123871.emapper.annotations\n",
      "INFO:planter.database.builder:Processed ERR9123871: 30281 sequences, 27884 annotations, 0 duplicates\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>status</th>\n",
       "      <th>sequences_loaded</th>\n",
       "      <th>annotations_loaded</th>\n",
       "      <th>duplicates</th>\n",
       "      <th>error</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR9123871</td>\n",
       "      <td>success</td>\n",
       "      <td>30281</td>\n",
       "      <td>27884</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_id   status  sequences_loaded  annotations_loaded  duplicates error\n",
       "0  ERR9123871  success             30281               27884           0  None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total_sequences</th>\n",
       "      <th>total_samples</th>\n",
       "      <th>representative_sequences</th>\n",
       "      <th>annotated_sequences</th>\n",
       "      <th>sequences_with_go</th>\n",
       "      <th>sequences_with_ec</th>\n",
       "      <th>total_clusters</th>\n",
       "      <th>avg_sequence_length</th>\n",
       "      <th>min_sequence_length</th>\n",
       "      <th>max_sequence_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30281</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>27884</td>\n",
       "      <td>15140</td>\n",
       "      <td>6682</td>\n",
       "      <td>0</td>\n",
       "      <td>441.07</td>\n",
       "      <td>100</td>\n",
       "      <td>3243</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   total_sequences  total_samples  representative_sequences  \\\n",
       "0            30281              1                         0   \n",
       "\n",
       "   annotated_sequences  sequences_with_go  sequences_with_ec  total_clusters  \\\n",
       "0                27884              15140               6682               0   \n",
       "\n",
       "   avg_sequence_length  min_sequence_length  max_sequence_length  \n",
       "0               441.07                  100                 3243  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from planter.database.builder import SequenceDBBuilder\n",
    "# Define paths\n",
    "DB_PATH = \"/mnt/data2/planter_outputs/planter-test.duckdb\"\n",
    "OUTPUT_DIR = Path(\"/mnt/data2/planter_outputs\")\n",
    "# CLUSTER_TSV = \"/mnt/data/planter_outputs/clusters.tsv\"  # if you have clustering results\n",
    "\n",
    "# List of SRA IDs to process\n",
    "# sample_ids = ['ERR9123871', 'ERR9123872', 'ERR9123874', 'ERR9123875', 'ERR9123876']\n",
    "\n",
    "# sample_ids = [\n",
    "#     'ERR9123871', 'ERR9123872', 'ERR9123874', 'ERR9123875', 'ERR9123876', \n",
    "#     'ERR9123877', 'ERR9123878', 'ERR9123879', 'ERR9123880', 'ERR9123881', \n",
    "#     'ERR9123882', 'SRR10444679', 'SRR10444680', 'SRR10444681', 'SRR10444682', \n",
    "#     'SRR10444683', 'SRR10444684', 'SRR11011255', 'SRR11011256', 'SRR11011257', \n",
    "#     'SRR11011258', 'SRR11011259', 'SRR11011260', 'SRR12068547', 'SRR128113', \n",
    "#     'SRR128114', 'SRR13765006', 'SRR14292007', 'SRR14292008', 'SRR18070778', \n",
    "#     'SRR18070779', 'SRR18070780', 'SRR18070781', 'SRR18070782', 'SRR18070783', \n",
    "#     'SRR18070784', 'SRR18070785', 'SRR18070786', 'SRR18070787', 'SRR18070788', \n",
    "#     'SRR18070789', 'SRR18070790', 'SRR18070791', 'SRR18070792', 'SRR18070793', \n",
    "#     'SRR18070794', 'SRR18070795', 'SRR18735292', 'SRR19034772', 'SRR19034773', \n",
    "#     'SRR19619612', 'SRR19619613', 'SRR19619614', 'SRR22271585', 'SRR22271586', \n",
    "#     'SRR22271587', 'SRR22271588', 'SRR22271589', 'SRR22904707', 'SRR24974225', \n",
    "#     'SRR24974226', 'SRR24974227', 'SRR24974228', 'SRR25582085', 'SRR29366264', \n",
    "#     'SRR29366265', 'SRR29366266', 'SRR5489198', 'SRR5992919', 'SRR5992920', \n",
    "#     'SRR6048009', 'SRR8859643', 'SRR8859644', 'SRR8859645', \n",
    "#     'SRR8859646', 'SRR8859647', 'SRR8859648'\n",
    "# ]\n",
    "\n",
    "sample_ids = ['ERR9123871']\n",
    "# Build database\n",
    "with SequenceDBBuilder(DB_PATH, output_dir=OUTPUT_DIR) as builder:\n",
    "    # Build initial database with sequences and annotations\n",
    "    results = builder.build_database(sample_ids)\n",
    "    display(results)\n",
    "    \n",
    "    # # Optionally load clustering results if available\n",
    "    # if Path(CLUSTER_TSV).exists():\n",
    "    #     builder.load_clusters_from_tsv(CLUSTER_TSV)\n",
    "    \n",
    "    # Get final database summary\n",
    "    summary = builder.get_database_summary()\n",
    "    display(summary)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Query"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get the database summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total_sequences</th>\n",
       "      <th>total_samples</th>\n",
       "      <th>representative_sequences</th>\n",
       "      <th>annotated_sequences</th>\n",
       "      <th>sequences_with_go</th>\n",
       "      <th>sequences_with_ec</th>\n",
       "      <th>total_clusters</th>\n",
       "      <th>avg_sequence_length</th>\n",
       "      <th>min_sequence_length</th>\n",
       "      <th>max_sequence_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1985570</td>\n",
       "      <td>77</td>\n",
       "      <td>0</td>\n",
       "      <td>1637348</td>\n",
       "      <td>785398</td>\n",
       "      <td>387029</td>\n",
       "      <td>0</td>\n",
       "      <td>430.94</td>\n",
       "      <td>100</td>\n",
       "      <td>7993</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   total_sequences  total_samples  representative_sequences  \\\n",
       "0          1985570             77                         0   \n",
       "\n",
       "   annotated_sequences  sequences_with_go  sequences_with_ec  total_clusters  \\\n",
       "0              1637348             785398             387029               0   \n",
       "\n",
       "   avg_sequence_length  min_sequence_length  max_sequence_length  \n",
       "0               430.94                  100                 7993  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from planter.database.query_manager import DatabaseManager\n",
    "db_path = \"/mnt/data2/planter_outputs/planter2.duckdb\"\n",
    "\n",
    "with DatabaseManager(db_path) as db_manager:\n",
    "    summary = db_manager.query_manager.database_summary()\n",
    "    display(summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## See what tables and schemas are present in the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tables in database:\n",
      "              name\n",
      "0      annotations\n",
      "1  cluster_members\n",
      "2         clusters\n",
      "3       ec_numbers\n",
      "4         go_terms\n",
      "5        kegg_info\n",
      "6   schema_version\n",
      "7        sequences\n",
      "8     sra_metadata\n",
      "\n",
      "Schema for table: annotations\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>notnull</th>\n",
       "      <th>dflt_value</th>\n",
       "      <th>pk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>seqhash_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>seed_ortholog</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>evalue</td>\n",
       "      <td>DOUBLE</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>score</td>\n",
       "      <td>DOUBLE</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>eggnog_ogs</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>max_annot_lvl</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>cog_category</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>description</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>preferred_name</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>sample_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cid            name     type  notnull dflt_value     pk\n",
       "0    0      seqhash_id  VARCHAR     True       None   True\n",
       "1    1   seed_ortholog  VARCHAR    False       None  False\n",
       "2    2          evalue   DOUBLE    False       None  False\n",
       "3    3           score   DOUBLE    False       None  False\n",
       "4    4      eggnog_ogs  VARCHAR    False       None  False\n",
       "5    5   max_annot_lvl  VARCHAR    False       None  False\n",
       "6    6    cog_category  VARCHAR    False       None  False\n",
       "7    7     description  VARCHAR    False       None  False\n",
       "8    8  preferred_name  VARCHAR    False       None  False\n",
       "9    9       sample_id  VARCHAR     True       None  False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Schema for table: cluster_members\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>notnull</th>\n",
       "      <th>dflt_value</th>\n",
       "      <th>pk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>seqhash_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>cluster_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cid        name     type  notnull dflt_value     pk\n",
       "0    0  seqhash_id  VARCHAR     True       None   True\n",
       "1    1  cluster_id  VARCHAR     True       None  False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Schema for table: clusters\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>notnull</th>\n",
       "      <th>dflt_value</th>\n",
       "      <th>pk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>cluster_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>representative_seqhash_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>size</td>\n",
       "      <td>INTEGER</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cid                       name     type  notnull dflt_value     pk\n",
       "0    0                 cluster_id  VARCHAR     True       None   True\n",
       "1    1  representative_seqhash_id  VARCHAR     True       None  False\n",
       "2    2                       size  INTEGER     True       None  False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Schema for table: ec_numbers\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>notnull</th>\n",
       "      <th>dflt_value</th>\n",
       "      <th>pk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>seqhash_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ec_number</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cid        name     type  notnull dflt_value    pk\n",
       "0    0  seqhash_id  VARCHAR     True       None  True\n",
       "1    1   ec_number  VARCHAR     True       None  True"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Schema for table: go_terms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>notnull</th>\n",
       "      <th>dflt_value</th>\n",
       "      <th>pk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>seqhash_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>go_term</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cid        name     type  notnull dflt_value    pk\n",
       "0    0  seqhash_id  VARCHAR     True       None  True\n",
       "1    1     go_term  VARCHAR     True       None  True"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Schema for table: kegg_info\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>notnull</th>\n",
       "      <th>dflt_value</th>\n",
       "      <th>pk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>seqhash_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>kegg_ko</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>kegg_pathway</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>kegg_module</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>kegg_reaction</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>kegg_rclass</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cid           name     type  notnull dflt_value     pk\n",
       "0    0     seqhash_id  VARCHAR     True       None   True\n",
       "1    1        kegg_ko  VARCHAR    False       None  False\n",
       "2    2   kegg_pathway  VARCHAR    False       None  False\n",
       "3    3    kegg_module  VARCHAR    False       None  False\n",
       "4    4  kegg_reaction  VARCHAR    False       None  False\n",
       "5    5    kegg_rclass  VARCHAR    False       None  False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Schema for table: schema_version\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>notnull</th>\n",
       "      <th>dflt_value</th>\n",
       "      <th>pk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>version</td>\n",
       "      <td>INTEGER</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>migration_name</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>applied_at</td>\n",
       "      <td>TIMESTAMP</td>\n",
       "      <td>False</td>\n",
       "      <td>CURRENT_TIMESTAMP</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cid            name       type  notnull         dflt_value     pk\n",
       "0    0         version    INTEGER     True               None   True\n",
       "1    1  migration_name    VARCHAR     True               None  False\n",
       "2    2      applied_at  TIMESTAMP    False  CURRENT_TIMESTAMP  False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Schema for table: sequences\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>notnull</th>\n",
       "      <th>dflt_value</th>\n",
       "      <th>pk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>seqhash_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>sequence</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>sample_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>assembly_date</td>\n",
       "      <td>TIMESTAMP</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>is_representative</td>\n",
       "      <td>BOOLEAN</td>\n",
       "      <td>True</td>\n",
       "      <td>CAST('f' AS BOOLEAN)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>length</td>\n",
       "      <td>INTEGER</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cid               name       type  notnull            dflt_value     pk\n",
       "0    0         seqhash_id    VARCHAR     True                  None   True\n",
       "1    1           sequence    VARCHAR     True                  None  False\n",
       "2    2          sample_id    VARCHAR     True                  None  False\n",
       "3    3      assembly_date  TIMESTAMP     True                  None  False\n",
       "4    4  is_representative    BOOLEAN     True  CAST('f' AS BOOLEAN)  False\n",
       "5    5             length    INTEGER     True                  None  False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Schema for table: sra_metadata\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>notnull</th>\n",
       "      <th>dflt_value</th>\n",
       "      <th>pk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>sample_id</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>organism</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>study_title</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>study_abstract</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>bioproject</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>biosample</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>library_strategy</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>library_source</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>library_selection</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>library_layout</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>instrument</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>run_spots</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>run_bases</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>run_published</td>\n",
       "      <td>VARCHAR</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cid               name     type  notnull dflt_value     pk\n",
       "0     0          sample_id  VARCHAR     True       None   True\n",
       "1     1           organism  VARCHAR    False       None  False\n",
       "2     2        study_title  VARCHAR    False       None  False\n",
       "3     3     study_abstract  VARCHAR    False       None  False\n",
       "4     4         bioproject  VARCHAR    False       None  False\n",
       "5     5          biosample  VARCHAR    False       None  False\n",
       "6     6   library_strategy  VARCHAR    False       None  False\n",
       "7     7     library_source  VARCHAR    False       None  False\n",
       "8     8  library_selection  VARCHAR    False       None  False\n",
       "9     9     library_layout  VARCHAR    False       None  False\n",
       "10   10         instrument  VARCHAR    False       None  False\n",
       "11   11          run_spots  VARCHAR    False       None  False\n",
       "12   12          run_bases  VARCHAR    False       None  False\n",
       "13   13      run_published  VARCHAR    False       None  False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with DatabaseManager(db_path) as db_manager:\n",
    "    # Fetch all tables\n",
    "    tables = db_manager.con.execute(\"SHOW TABLES;\").fetchdf()\n",
    "    print(\"Tables in database:\")\n",
    "    print(tables)\n",
    "\n",
    "    # Iterate through each table and fetch its schema\n",
    "    for table in tables['name']:\n",
    "        print(f\"\\nSchema for table: {table}\")\n",
    "        schema = db_manager.con.execute(f\"PRAGMA table_info('{table}');\").fetchdf()\n",
    "        display(schema)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get sequence_annotations for seqhashIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nAnnotations:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>description</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>cog_category</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>ec_numbers</th>\n",
       "      <th>kegg_ko</th>\n",
       "      <th>kegg_pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_633f05f5d947805a8eae5b38eea7cdd090f48c3...</td>\n",
       "      <td>SRR10444679</td>\n",
       "      <td>Xanthoria parietina</td>\n",
       "      <td>PKS_DH</td>\n",
       "      <td>-</td>\n",
       "      <td>Q</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_8813cfa2d04ae4cf4c316e335a01c9d81b66255...</td>\n",
       "      <td>SRR128114</td>\n",
       "      <td>Digitalis purpurea</td>\n",
       "      <td>Belongs to the UDP-glycosyltransferase family</td>\n",
       "      <td>-</td>\n",
       "      <td>CG</td>\n",
       "      <td>None</td>\n",
       "      <td>2.4.1.210</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_416333688207b36e52865182ab219b807e55f9b...</td>\n",
       "      <td>SRR24974225</td>\n",
       "      <td>Digitalis purpurea</td>\n",
       "      <td>Belongs to the UDP-glycosyltransferase family</td>\n",
       "      <td>-</td>\n",
       "      <td>CG</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id    sample_id  \\\n",
       "0  v1_DLS_633f05f5d947805a8eae5b38eea7cdd090f48c3...  SRR10444679   \n",
       "1  v1_DLS_8813cfa2d04ae4cf4c316e335a01c9d81b66255...    SRR128114   \n",
       "2  v1_DLS_416333688207b36e52865182ab219b807e55f9b...  SRR24974225   \n",
       "\n",
       "              organism                                    description  \\\n",
       "0  Xanthoria parietina                                         PKS_DH   \n",
       "1   Digitalis purpurea  Belongs to the UDP-glycosyltransferase family   \n",
       "2   Digitalis purpurea  Belongs to the UDP-glycosyltransferase family   \n",
       "\n",
       "  preferred_name cog_category go_terms ec_numbers kegg_ko kegg_pathway  \n",
       "0              -            Q     None       None    None         None  \n",
       "1              -           CG     None  2.4.1.210    None         None  \n",
       "2              -           CG     None       None    None         None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from planter.database.query_manager import DatabaseManager\n",
    "import pandas as pd\n",
    "\n",
    "db_path = \"/mnt/data2/planter_outputs/planter2.duckdb\"\n",
    "\n",
    "# Test some seqhash IDs from your MMseqs results\n",
    "test_ids = [\n",
    "   \"v1_DLS_8813cfa2d04ae4cf4c316e335a01c9d81b66255681bc3e51a98cd00dc5563466.p2\",\n",
    "   \"v1_DLS_416333688207b36e52865182ab219b807e55f9b9cc56f30bb249fd4faa38e3da.p1\",\n",
    "   \"v1_DLS_9ec0dd0c4e615b3c23b1b8e475cd0dc87f96b87d66fb13d2d033dc87dc652ca1.p1\",\n",
    "   \"v1_DLS_633f05f5d947805a8eae5b38eea7cdd090f48c3dc1e3ce89eb0b6732042591aa.p1\",\n",
    "   \"v1_DLS_700011b382d7c61a71117db9be1cfde6ad9435a22f74e60d769f57737d13c55b.p1\"\n",
    "]\n",
    "\n",
    "with DatabaseManager(db_path) as db:\n",
    "    annotations = db.query_manager.sequence_annotations(values=(test_ids,))\n",
    "    display(\"\\nAnnotations:\", annotations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Search sequences\n",
    "\n",
    "You can parameterize the search.\n",
    "\n",
    "Examples show filtering by sample ID, length, GO Term, and COG category."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>length</th>\n",
       "      <th>is_representative</th>\n",
       "      <th>description</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>cog_category</th>\n",
       "      <th>evalue</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>cluster_size</th>\n",
       "      <th>organism</th>\n",
       "      <th>study_title</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_878049f1c7d690c2a475148c24aa6215193325f...</td>\n",
       "      <td>SRR5992919</td>\n",
       "      <td>367</td>\n",
       "      <td>False</td>\n",
       "      <td>U3 snoRNA binding</td>\n",
       "      <td>BMS1</td>\n",
       "      <td>J</td>\n",
       "      <td>3.060000e-13</td>\n",
       "      <td>653948.CCA25717</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Zophobas atratus</td>\n",
       "      <td>De novo transcriptome assembly of the Malpighi...</td>\n",
       "      <td>PRJNA400859</td>\n",
       "      <td>SAMN07581518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_fc5046ddce2a1362613dcc9c97083bfd3e89179...</td>\n",
       "      <td>SRR10444679</td>\n",
       "      <td>502</td>\n",
       "      <td>False</td>\n",
       "      <td>structural molecule activity</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>1.340000e-08</td>\n",
       "      <td>122962.PRO_0000039759</td>\n",
       "      <td>464095|Picornavirales</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xanthoria parietina</td>\n",
       "      <td>Xanthoria parietina 46-1 Gene Expression Profi...</td>\n",
       "      <td>PRJNA584076</td>\n",
       "      <td>SAMN13173247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_e5f14c12609ce0b15d873580594e6e6aa4766eb...</td>\n",
       "      <td>SRR5992920</td>\n",
       "      <td>110</td>\n",
       "      <td>False</td>\n",
       "      <td>Ribosome biogenesis protein (BMS1</td>\n",
       "      <td>-</td>\n",
       "      <td>J</td>\n",
       "      <td>1.290000e-20</td>\n",
       "      <td>110365.A0A023B5C2</td>\n",
       "      <td>5794|Apicomplexa</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Zophobas atratus</td>\n",
       "      <td>De novo transcriptome assembly of the Malpighi...</td>\n",
       "      <td>PRJNA400859</td>\n",
       "      <td>SAMN07581519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>v1_DLS_81d3635e002a71c427b5ec350907c012f8dcd00...</td>\n",
       "      <td>SRR10444679</td>\n",
       "      <td>491</td>\n",
       "      <td>False</td>\n",
       "      <td>structural molecule activity</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>2.550000e-33</td>\n",
       "      <td>928300.PRO_0000398370</td>\n",
       "      <td>464095|Picornavirales</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xanthoria parietina</td>\n",
       "      <td>Xanthoria parietina 46-1 Gene Expression Profi...</td>\n",
       "      <td>PRJNA584076</td>\n",
       "      <td>SAMN13173247</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id    sample_id  length  \\\n",
       "0  v1_DLS_878049f1c7d690c2a475148c24aa6215193325f...   SRR5992919     367   \n",
       "1  v1_DLS_fc5046ddce2a1362613dcc9c97083bfd3e89179...  SRR10444679     502   \n",
       "2  v1_DLS_e5f14c12609ce0b15d873580594e6e6aa4766eb...   SRR5992920     110   \n",
       "3  v1_DLS_81d3635e002a71c427b5ec350907c012f8dcd00...  SRR10444679     491   \n",
       "\n",
       "   is_representative                        description preferred_name  \\\n",
       "0              False                  U3 snoRNA binding           BMS1   \n",
       "1              False       structural molecule activity              -   \n",
       "2              False  Ribosome biogenesis protein (BMS1              -   \n",
       "3              False       structural molecule activity              -   \n",
       "\n",
       "  cog_category        evalue          seed_ortholog          max_annot_lvl  \\\n",
       "0            J  3.060000e-13        653948.CCA25717         2759|Eukaryota   \n",
       "1            A  1.340000e-08  122962.PRO_0000039759  464095|Picornavirales   \n",
       "2            J  1.290000e-20      110365.A0A023B5C2       5794|Apicomplexa   \n",
       "3            A  2.550000e-33  928300.PRO_0000398370  464095|Picornavirales   \n",
       "\n",
       "  cluster_id  cluster_size             organism  \\\n",
       "0       None           NaN     Zophobas atratus   \n",
       "1       None           NaN  Xanthoria parietina   \n",
       "2       None           NaN     Zophobas atratus   \n",
       "3       None           NaN  Xanthoria parietina   \n",
       "\n",
       "                                         study_title   bioproject  \\\n",
       "0  De novo transcriptome assembly of the Malpighi...  PRJNA400859   \n",
       "1  Xanthoria parietina 46-1 Gene Expression Profi...  PRJNA584076   \n",
       "2  De novo transcriptome assembly of the Malpighi...  PRJNA400859   \n",
       "3  Xanthoria parietina 46-1 Gene Expression Profi...  PRJNA584076   \n",
       "\n",
       "      biosample  \n",
       "0  SAMN07581518  \n",
       "1  SAMN13173247  \n",
       "2  SAMN07581519  \n",
       "3  SAMN13173247  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from planter.database.query_manager import DatabaseManager\n",
    "\n",
    "db_path = \"/mnt/data2/planter_outputs/planter2.duckdb\"\n",
    "output_dir = \"/mnt/data2/planter_outputs\"\n",
    "\n",
    "with DatabaseManager(db_path) as db_manager:\n",
    "    sample_ids = [\"SRR18070780\", \"SRR10444679\"]\n",
    "    cog_categories = ['Q']  # desired COG categories\n",
    "    go_terms = ['GO:0075109']  # example GO terms to search\n",
    "\n",
    "    params = {\n",
    "        # \"sample_id_condition\": sample_ids,  # e.g., \"s.sample_id IN ('SRR18070780', 'SRR18070781')\"\n",
    "        # \"cog_category_condition\": ['M', 'Q'],\n",
    "        \"go_term_condition\": go_terms,  # pass a list of GO terms        \n",
    "        \"min_length_condition\": \"s.length >= 100\",\n",
    "        \"max_length_condition\": \"s.length <= 1000\",\n",
    "        \"description_condition\": None,\n",
    "        \"organism_condition\": None,\n",
    "        # \"limit\": 10,\n",
    "    }\n",
    "\n",
    "    results = db_manager.query_manager.search_sequences(params=params)\n",
    "    display(results)\n",
    "\n",
    "    # call go_term_summary dynamically\n",
    "    # go_summary = db_manager.query_manager.go_term_summary(None, None, 5)\n",
    "    # print(\"\\nGO term summary:\")\n",
    "    # display(go_summary)\n",
    "\n",
    "    # call organism_summary dynamically\n",
    "    # organism_summary = db_manager.query_manager.organism_summary()\n",
    "    # print(\"\\nOrganism summary:\")\n",
    "    # display(organism_summary)\n",
    "    # print(dir(db_manager.query_manager))\n",
    "    # db_manager.query_manager.search_sequences('v1_DLS_31412ec4347c212e6892097053de8dc39cd53341988080fd7b80866c35840a0a.p1')\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add new pipeline result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:planter.database.builder:Fetching metadata for ERR4235136\n",
      "INFO:planter.database.utils.sra:Fetching metadata for ERR4235136\n",
      "INFO:planter.database.builder:Loading sequences from /mnt/data2/planter_outputs/ERR4235136/transdecoder/ERR4235136.pep\n",
      "INFO:planter.database.builder:Loaded 42569 new sequences for ERR4235136\n",
      "INFO:planter.database.builder:Loading annotations from /mnt/data2/planter_outputs/ERR4235136/eggnog/ERR4235136.emapper.annotations\n",
      "INFO:planter.database.builder:Added ERR4235136: 42569 sequences, 31086 annotations\n"
     ]
    }
   ],
   "source": [
    "from planter.database.builder import SequenceDBBuilder\n",
    "# Define paths\n",
    "DB_PATH = \"/mnt/data2/planter_outputs/planter-test.duckdb\"\n",
    "OUTPUT_DIR = Path(\"/mnt/data2/planter_outputs\")\n",
    "\n",
    "with SequenceDBBuilder(DB_PATH, output_dir=OUTPUT_DIR) as builder:\n",
    "    # Add new sample\n",
    "    new_results = builder.update_database(['ERR4235136'])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
