{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "from IPython.display import display\n",
    "\n",
    "import duckdb\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import logging\n",
    "from Bio import SeqIO\n",
    "from datetime import datetime\n",
    "import time\n",
    "\n",
    "from typing import List, Optional, Union, Dict, Set\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "logger = logging.getLogger(__name__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of annotations:\n",
      "annotations has 251 rows and 10 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggnog_ogs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>cog_category</th>\n",
       "      <th>description</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>265311.Mfl224</td>\n",
       "      <td>1.930000e-06</td>\n",
       "      <td>53.5</td>\n",
       "      <td>COG1564@1|root,COG1564@2|Bacteria,3WTAN@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>H</td>\n",
       "      <td>Thiamin pyrophosphokinase, vitamin B1 binding ...</td>\n",
       "      <td>thiN</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>265311.Mfl040</td>\n",
       "      <td>3.740000e-45</td>\n",
       "      <td>155.0</td>\n",
       "      <td>COG0022@1|root,COG0022@2|Bacteria,3WSYT@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>C</td>\n",
       "      <td>dehydrogenase e1 component</td>\n",
       "      <td>pdhB</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...</td>\n",
       "      <td>265311.Mfl489</td>\n",
       "      <td>1.300000e-29</td>\n",
       "      <td>120.0</td>\n",
       "      <td>2EWBJ@1|root,33PQA@2|Bacteria,3WT4V@544448|Ten...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  seed_ortholog  \\\n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...  265311.Mfl224   \n",
       "1  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...  265311.Mfl040   \n",
       "2  v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...  265311.Mfl489   \n",
       "\n",
       "         evalue  score                                         eggnog_ogs  \\\n",
       "0  1.930000e-06   53.5  COG1564@1|root,COG1564@2|Bacteria,3WTAN@544448...   \n",
       "1  3.740000e-45  155.0  COG0022@1|root,COG0022@2|Bacteria,3WSYT@544448...   \n",
       "2  1.300000e-29  120.0  2EWBJ@1|root,33PQA@2|Bacteria,3WT4V@544448|Ten...   \n",
       "\n",
       "        max_annot_lvl cog_category  \\\n",
       "0  544448|Tenericutes            H   \n",
       "1  544448|Tenericutes            C   \n",
       "2  544448|Tenericutes            -   \n",
       "\n",
       "                                         description preferred_name  \\\n",
       "0  Thiamin pyrophosphokinase, vitamin B1 binding ...           thiN   \n",
       "1                         dehydrogenase e1 component           pdhB   \n",
       "2                                                  -              -   \n",
       "\n",
       "     sample_id  \n",
       "0  SRR12068547  \n",
       "1  SRR12068547  \n",
       "2  SRR12068547  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of cluster_members:\n",
      "cluster_members has 0 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>cluster_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, cluster_id]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of clusters:\n",
      "clusters has 0 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>representative_seqhash_id</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cluster_id, representative_seqhash_id, size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of ec_numbers:\n",
      "ec_numbers has 130 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_05fc6dc2fcc79581a85e20a59498e2ca4389b43...</td>\n",
       "      <td>1.2.4.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_116aa24751cc0c03d972d124b07632ab5a2db0e...</td>\n",
       "      <td>3.4.23.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_166a64b63f322440c5a41bc6219f0a942373c2b...</td>\n",
       "      <td>5.4.2.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  ec_number\n",
       "0  v1_DLS_05fc6dc2fcc79581a85e20a59498e2ca4389b43...    1.2.4.4\n",
       "1  v1_DLS_116aa24751cc0c03d972d124b07632ab5a2db0e...  3.4.23.36\n",
       "2  v1_DLS_166a64b63f322440c5a41bc6219f0a942373c2b...   5.4.2.12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of go_terms:\n",
      "go_terms has 110 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>go_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>GO:0140032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_1016173829ac36ac2333823a671a085347bddb2...</td>\n",
       "      <td>GO:0005515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_27169752765a0e3a22d9727f0dcb6ece7c520a9...</td>\n",
       "      <td>GO:0005622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id     go_term\n",
       "0  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...  GO:0140032\n",
       "1  v1_DLS_1016173829ac36ac2333823a671a085347bddb2...  GO:0005515\n",
       "2  v1_DLS_27169752765a0e3a22d9727f0dcb6ece7c520a9...  GO:0005622"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of kegg_info:\n",
      "kegg_info has 0 rows and 6 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>kegg_ko</th>\n",
       "      <th>kegg_pathway</th>\n",
       "      <th>kegg_module</th>\n",
       "      <th>kegg_reaction</th>\n",
       "      <th>kegg_rclass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, kegg_ko, kegg_pathway, kegg_module, kegg_reaction, kegg_rclass]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of schema_version:\n",
      "schema_version has 3 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>version</th>\n",
       "      <th>migration_name</th>\n",
       "      <th>applied_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>000_schema_version.sql</td>\n",
       "      <td>2025-02-11 02:04:08.915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>001_initial_schema.sql</td>\n",
       "      <td>2025-02-11 02:04:08.915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>002_add_indexes.sql</td>\n",
       "      <td>2025-02-11 02:04:08.915</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   version          migration_name              applied_at\n",
       "0        1  000_schema_version.sql 2025-02-11 02:04:08.915\n",
       "1        2  001_initial_schema.sql 2025-02-11 02:04:08.915\n",
       "2        3     002_add_indexes.sql 2025-02-11 02:04:08.915"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sequences:\n",
      "sequences has 437 rows and 7 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>assembly_date</th>\n",
       "      <th>is_representative</th>\n",
       "      <th>repseq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774568</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...</td>\n",
       "      <td>MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774589</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774599</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  \\\n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...   \n",
       "1  v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...   \n",
       "2  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...   \n",
       "\n",
       "                                            sequence    sample_id  \\\n",
       "0  MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  SRR12068547   \n",
       "1  MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  SRR12068547   \n",
       "2  MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...  SRR12068547   \n",
       "\n",
       "               assembly_date  is_representative  \\\n",
       "0 2025-02-11 02:04:12.774568              False   \n",
       "1 2025-02-11 02:04:12.774589              False   \n",
       "2 2025-02-11 02:04:12.774599              False   \n",
       "\n",
       "                                           repseq_id  length  \n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...     160  \n",
       "1  v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...     229  \n",
       "2  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...     117  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sra_metadata:\n",
      "sra_metadata has 1 rows and 14 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_abstract</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>instrument</th>\n",
       "      <th>run_spots</th>\n",
       "      <th>run_bases</th>\n",
       "      <th>run_published</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>Transcriptome profiling of Mesoplasma florum s...</td>\n",
       "      <td>Mesoplasma florum is a near-minimal bacterium ...</td>\n",
       "      <td>PRJNA641121</td>\n",
       "      <td>SAMN15341344</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id           organism  \\\n",
       "0  SRR12068547  Mesoplasma florum   \n",
       "\n",
       "                                         study_title  \\\n",
       "0  Transcriptome profiling of Mesoplasma florum s...   \n",
       "\n",
       "                                      study_abstract   bioproject  \\\n",
       "0  Mesoplasma florum is a near-minimal bacterium ...  PRJNA641121   \n",
       "\n",
       "      biosample library_strategy library_source library_selection  \\\n",
       "0  SAMN15341344             None           None              None   \n",
       "\n",
       "  library_layout instrument run_spots run_bases run_published  \n",
       "0           None   ILLUMINA      None      None          None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of annotations:\n",
      "annotations has 320 rows and 10 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggnog_ogs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>cog_category</th>\n",
       "      <th>description</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0045223251933025812ef4b1f3f2c589a13ffa3...</td>\n",
       "      <td>265311.Mfl097</td>\n",
       "      <td>3.620000e-93</td>\n",
       "      <td>288.0</td>\n",
       "      <td>COG0444@1|root,COG4608@1|root,COG0444@2|Bacter...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>P</td>\n",
       "      <td>Belongs to the ABC transporter superfamily</td>\n",
       "      <td>oppF</td>\n",
       "      <td>SRR12068550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_009e48be93dbf3e6aa24c3efcc2cd6582490192...</td>\n",
       "      <td>265311.Mfl111</td>\n",
       "      <td>2.390000e-113</td>\n",
       "      <td>326.0</td>\n",
       "      <td>COG0711@1|root,COG0711@2|Bacteria,3WTD0@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>C</td>\n",
       "      <td>Component of the F(0) channel, it forms part o...</td>\n",
       "      <td>atpF</td>\n",
       "      <td>SRR12068550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_00c863de08751623595b1ec55aa36ce3cd37216...</td>\n",
       "      <td>265311.Mfl111</td>\n",
       "      <td>2.390000e-113</td>\n",
       "      <td>326.0</td>\n",
       "      <td>COG0711@1|root,COG0711@2|Bacteria,3WTD0@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>C</td>\n",
       "      <td>Component of the F(0) channel, it forms part o...</td>\n",
       "      <td>atpF</td>\n",
       "      <td>SRR12068550</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  seed_ortholog  \\\n",
       "0  v1_DLS_0045223251933025812ef4b1f3f2c589a13ffa3...  265311.Mfl097   \n",
       "1  v1_DLS_009e48be93dbf3e6aa24c3efcc2cd6582490192...  265311.Mfl111   \n",
       "2  v1_DLS_00c863de08751623595b1ec55aa36ce3cd37216...  265311.Mfl111   \n",
       "\n",
       "          evalue  score                                         eggnog_ogs  \\\n",
       "0   3.620000e-93  288.0  COG0444@1|root,COG4608@1|root,COG0444@2|Bacter...   \n",
       "1  2.390000e-113  326.0  COG0711@1|root,COG0711@2|Bacteria,3WTD0@544448...   \n",
       "2  2.390000e-113  326.0  COG0711@1|root,COG0711@2|Bacteria,3WTD0@544448...   \n",
       "\n",
       "        max_annot_lvl cog_category  \\\n",
       "0  544448|Tenericutes            P   \n",
       "1  544448|Tenericutes            C   \n",
       "2  544448|Tenericutes            C   \n",
       "\n",
       "                                         description preferred_name  \\\n",
       "0         Belongs to the ABC transporter superfamily           oppF   \n",
       "1  Component of the F(0) channel, it forms part o...           atpF   \n",
       "2  Component of the F(0) channel, it forms part o...           atpF   \n",
       "\n",
       "     sample_id  \n",
       "0  SRR12068550  \n",
       "1  SRR12068550  \n",
       "2  SRR12068550  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of cluster_members:\n",
      "cluster_members has 0 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>cluster_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, cluster_id]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of clusters:\n",
      "clusters has 0 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>representative_seqhash_id</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cluster_id, representative_seqhash_id, size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of ec_numbers:\n",
      "ec_numbers has 144 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0c8eb4715304e15ccd1af0a087202a1e20b3c28...</td>\n",
       "      <td>5.99.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_119a5ee2390475746fff8bdab1c8c9a5d6bdc03...</td>\n",
       "      <td>2.7.1.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_1717e40d619e5ed64ed6baf2d7c5de070f0c551...</td>\n",
       "      <td>2.7.1.76</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id ec_number\n",
       "0  v1_DLS_0c8eb4715304e15ccd1af0a087202a1e20b3c28...  5.99.1.3\n",
       "1  v1_DLS_119a5ee2390475746fff8bdab1c8c9a5d6bdc03...  2.7.1.76\n",
       "2  v1_DLS_1717e40d619e5ed64ed6baf2d7c5de070f0c551...  2.7.1.76"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of go_terms:\n",
      "go_terms has 214 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>go_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_203caeecec00ae2e766e821b8d4b5aac54085dc...</td>\n",
       "      <td>GO:0005575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_4082c0cc7a2f556d8672eb7d893e708426e0dd4...</td>\n",
       "      <td>GO:0005575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_5ac4e49f5163173bbecbc5e8047e483c30ccb2d...</td>\n",
       "      <td>GO:0044237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id     go_term\n",
       "0  v1_DLS_203caeecec00ae2e766e821b8d4b5aac54085dc...  GO:0005575\n",
       "1  v1_DLS_4082c0cc7a2f556d8672eb7d893e708426e0dd4...  GO:0005575\n",
       "2  v1_DLS_5ac4e49f5163173bbecbc5e8047e483c30ccb2d...  GO:0044237"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of kegg_info:\n",
      "kegg_info has 0 rows and 6 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>kegg_ko</th>\n",
       "      <th>kegg_pathway</th>\n",
       "      <th>kegg_module</th>\n",
       "      <th>kegg_reaction</th>\n",
       "      <th>kegg_rclass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, kegg_ko, kegg_pathway, kegg_module, kegg_reaction, kegg_rclass]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of schema_version:\n",
      "schema_version has 3 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>version</th>\n",
       "      <th>migration_name</th>\n",
       "      <th>applied_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>000_schema_version.sql</td>\n",
       "      <td>2025-02-11 02:04:08.913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>001_initial_schema.sql</td>\n",
       "      <td>2025-02-11 02:04:08.913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>002_add_indexes.sql</td>\n",
       "      <td>2025-02-11 02:04:08.913</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   version          migration_name              applied_at\n",
       "0        1  000_schema_version.sql 2025-02-11 02:04:08.913\n",
       "1        2  001_initial_schema.sql 2025-02-11 02:04:08.913\n",
       "2        3     002_add_indexes.sql 2025-02-11 02:04:08.913"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sequences:\n",
      "sequences has 519 rows and 7 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>assembly_date</th>\n",
       "      <th>is_representative</th>\n",
       "      <th>repseq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0045223251933025812ef4b1f3f2c589a13ffa3...</td>\n",
       "      <td>MLPEHLSRYPHEFSGGQRQRIGIARALVMKPSFVVCDEPISALDVS...</td>\n",
       "      <td>SRR12068550</td>\n",
       "      <td>2025-02-11 02:04:10.514353</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_0045223251933025812ef4b1f3f2c589a13ffa3...</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_009e48be93dbf3e6aa24c3efcc2cd6582490192...</td>\n",
       "      <td>MIFFAETQTAGVPEIITSLFPNLPNFIAHVIATIVLVVILSKLMYK...</td>\n",
       "      <td>SRR12068550</td>\n",
       "      <td>2025-02-11 02:04:10.514371</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_009e48be93dbf3e6aa24c3efcc2cd6582490192...</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_00c863de08751623595b1ec55aa36ce3cd37216...</td>\n",
       "      <td>MNNYLXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "      <td>SRR12068550</td>\n",
       "      <td>2025-02-11 02:04:10.514381</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_00c863de08751623595b1ec55aa36ce3cd37216...</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  \\\n",
       "0  v1_DLS_0045223251933025812ef4b1f3f2c589a13ffa3...   \n",
       "1  v1_DLS_009e48be93dbf3e6aa24c3efcc2cd6582490192...   \n",
       "2  v1_DLS_00c863de08751623595b1ec55aa36ce3cd37216...   \n",
       "\n",
       "                                            sequence    sample_id  \\\n",
       "0  MLPEHLSRYPHEFSGGQRQRIGIARALVMKPSFVVCDEPISALDVS...  SRR12068550   \n",
       "1  MIFFAETQTAGVPEIITSLFPNLPNFIAHVIATIVLVVILSKLMYK...  SRR12068550   \n",
       "2  MNNYLXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  SRR12068550   \n",
       "\n",
       "               assembly_date  is_representative  \\\n",
       "0 2025-02-11 02:04:10.514353              False   \n",
       "1 2025-02-11 02:04:10.514371              False   \n",
       "2 2025-02-11 02:04:10.514381              False   \n",
       "\n",
       "                                           repseq_id  length  \n",
       "0  v1_DLS_0045223251933025812ef4b1f3f2c589a13ffa3...     148  \n",
       "1  v1_DLS_009e48be93dbf3e6aa24c3efcc2cd6582490192...     178  \n",
       "2  v1_DLS_00c863de08751623595b1ec55aa36ce3cd37216...     150  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sra_metadata:\n",
      "sra_metadata has 1 rows and 14 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_abstract</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>instrument</th>\n",
       "      <th>run_spots</th>\n",
       "      <th>run_bases</th>\n",
       "      <th>run_published</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR12068550</td>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>Transcriptome profiling of Mesoplasma florum s...</td>\n",
       "      <td>Mesoplasma florum is a near-minimal bacterium ...</td>\n",
       "      <td>PRJNA641121</td>\n",
       "      <td>SAMN15341341</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id           organism  \\\n",
       "0  SRR12068550  Mesoplasma florum   \n",
       "\n",
       "                                         study_title  \\\n",
       "0  Transcriptome profiling of Mesoplasma florum s...   \n",
       "\n",
       "                                      study_abstract   bioproject  \\\n",
       "0  Mesoplasma florum is a near-minimal bacterium ...  PRJNA641121   \n",
       "\n",
       "      biosample library_strategy library_source library_selection  \\\n",
       "0  SAMN15341341             None           None              None   \n",
       "\n",
       "  library_layout instrument run_spots run_bases run_published  \n",
       "0           None   ILLUMINA      None      None          None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "SAMPLES = ['SRR12068547', 'SRR12068550']\n",
    "DUCKDB_PATHS = [\n",
    "    Path('/mnt/data3/planter_outputs/SRR12068547/SRR12068547.duckdb'),\n",
    "    Path('/mnt/data3/planter_outputs/SRR12068550/SRR12068550.duckdb')\n",
    "]\n",
    "\n",
    "def display_tables(duckdb_path):\n",
    "    if not Path(duckdb_path).exists():\n",
    "        raise FileNotFoundError(f'{duckdb_path} does not exist')\n",
    "    with duckdb.connect(duckdb_path) as conn:\n",
    "        db = conn.execute(\"PRAGMA show_tables;\").fetchdf()\n",
    "        tables = db['name'].tolist()\n",
    "\n",
    "        for table in tables:\n",
    "            print(f\"First 3 rows of {table}:\")\n",
    "            df = conn.execute(f\"SELECT * FROM {table}\").fetchdf()\n",
    "            print(f'{table} has {df.shape[0]} rows and {df.shape[1]} columns')\n",
    "            display(df.head(3))\n",
    "\n",
    "display_tables(DUCKDB_PATHS[0])\n",
    "display_tables(DUCKDB_PATHS[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attaching /mnt/data3/planter_outputs/SRR12068547/SRR12068547.duckdb as db0...\n",
      "Finished merging /mnt/data3/planter_outputs/SRR12068547/SRR12068547.duckdb\n",
      "\n",
      "Attaching /mnt/data3/planter_outputs/SRR12068550/SRR12068550.duckdb as db1...\n",
      "Finished merging /mnt/data3/planter_outputs/SRR12068550/SRR12068550.duckdb\n",
      "\n",
      "All databases have been merged into: /tmp/merged.duckdb\n",
      "First 3 rows of annotations:\n",
      "annotations has 44023 rows and 10 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggnog_ogs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>cog_category</th>\n",
       "      <th>description</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>265311.Mfl224</td>\n",
       "      <td>1.930000e-06</td>\n",
       "      <td>53.5</td>\n",
       "      <td>COG1564@1|root,COG1564@2|Bacteria,3WTAN@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>H</td>\n",
       "      <td>Thiamin pyrophosphokinase, vitamin B1 binding ...</td>\n",
       "      <td>thiN</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>265311.Mfl040</td>\n",
       "      <td>3.740000e-45</td>\n",
       "      <td>155.0</td>\n",
       "      <td>COG0022@1|root,COG0022@2|Bacteria,3WSYT@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>C</td>\n",
       "      <td>dehydrogenase e1 component</td>\n",
       "      <td>pdhB</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...</td>\n",
       "      <td>265311.Mfl489</td>\n",
       "      <td>1.300000e-29</td>\n",
       "      <td>120.0</td>\n",
       "      <td>2EWBJ@1|root,33PQA@2|Bacteria,3WT4V@544448|Ten...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  seed_ortholog  \\\n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...  265311.Mfl224   \n",
       "1  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...  265311.Mfl040   \n",
       "2  v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...  265311.Mfl489   \n",
       "\n",
       "         evalue  score                                         eggnog_ogs  \\\n",
       "0  1.930000e-06   53.5  COG1564@1|root,COG1564@2|Bacteria,3WTAN@544448...   \n",
       "1  3.740000e-45  155.0  COG0022@1|root,COG0022@2|Bacteria,3WSYT@544448...   \n",
       "2  1.300000e-29  120.0  2EWBJ@1|root,33PQA@2|Bacteria,3WT4V@544448|Ten...   \n",
       "\n",
       "        max_annot_lvl cog_category  \\\n",
       "0  544448|Tenericutes            H   \n",
       "1  544448|Tenericutes            C   \n",
       "2  544448|Tenericutes            -   \n",
       "\n",
       "                                         description preferred_name  \\\n",
       "0  Thiamin pyrophosphokinase, vitamin B1 binding ...           thiN   \n",
       "1                         dehydrogenase e1 component           pdhB   \n",
       "2                                                  -              -   \n",
       "\n",
       "     sample_id  \n",
       "0  SRR12068547  \n",
       "1  SRR12068547  \n",
       "2  SRR12068547  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of cluster_members:\n",
      "cluster_members has 0 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>cluster_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, cluster_id]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of clusters:\n",
      "clusters has 0 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>representative_seqhash_id</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cluster_id, representative_seqhash_id, size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of ec_numbers:\n",
      "ec_numbers has 16492 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_05fc6dc2fcc79581a85e20a59498e2ca4389b43...</td>\n",
       "      <td>1.2.4.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_116aa24751cc0c03d972d124b07632ab5a2db0e...</td>\n",
       "      <td>3.4.23.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_166a64b63f322440c5a41bc6219f0a942373c2b...</td>\n",
       "      <td>5.4.2.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  ec_number\n",
       "0  v1_DLS_05fc6dc2fcc79581a85e20a59498e2ca4389b43...    1.2.4.4\n",
       "1  v1_DLS_116aa24751cc0c03d972d124b07632ab5a2db0e...  3.4.23.36\n",
       "2  v1_DLS_166a64b63f322440c5a41bc6219f0a942373c2b...   5.4.2.12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of go_terms:\n",
      "go_terms has 1994038 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>go_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>GO:0140032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_1016173829ac36ac2333823a671a085347bddb2...</td>\n",
       "      <td>GO:0005515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_27169752765a0e3a22d9727f0dcb6ece7c520a9...</td>\n",
       "      <td>GO:0005622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id     go_term\n",
       "0  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...  GO:0140032\n",
       "1  v1_DLS_1016173829ac36ac2333823a671a085347bddb2...  GO:0005515\n",
       "2  v1_DLS_27169752765a0e3a22d9727f0dcb6ece7c520a9...  GO:0005622"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of kegg_info:\n",
      "kegg_info has 0 rows and 6 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>kegg_ko</th>\n",
       "      <th>kegg_pathway</th>\n",
       "      <th>kegg_module</th>\n",
       "      <th>kegg_reaction</th>\n",
       "      <th>kegg_rclass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, kegg_ko, kegg_pathway, kegg_module, kegg_reaction, kegg_rclass]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of schema_version:\n",
      "schema_version has 0 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>version</th>\n",
       "      <th>migration_name</th>\n",
       "      <th>applied_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [version, migration_name, applied_at]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sequences:\n",
      "sequences has 85232 rows and 7 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>assembly_date</th>\n",
       "      <th>is_representative</th>\n",
       "      <th>repseq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774568</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...</td>\n",
       "      <td>MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774589</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774599</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  \\\n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...   \n",
       "1  v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...   \n",
       "2  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...   \n",
       "\n",
       "                                            sequence    sample_id  \\\n",
       "0  MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  SRR12068547   \n",
       "1  MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  SRR12068547   \n",
       "2  MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...  SRR12068547   \n",
       "\n",
       "               assembly_date  is_representative  \\\n",
       "0 2025-02-11 02:04:12.774568              False   \n",
       "1 2025-02-11 02:04:12.774589              False   \n",
       "2 2025-02-11 02:04:12.774599              False   \n",
       "\n",
       "                                           repseq_id  length  \n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...     160  \n",
       "1  v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...     229  \n",
       "2  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...     117  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sra_metadata:\n",
      "sra_metadata has 8 rows and 14 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_abstract</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>instrument</th>\n",
       "      <th>run_spots</th>\n",
       "      <th>run_bases</th>\n",
       "      <th>run_published</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>Transcriptome profiling of Mesoplasma florum s...</td>\n",
       "      <td>Mesoplasma florum is a near-minimal bacterium ...</td>\n",
       "      <td>PRJNA641121</td>\n",
       "      <td>SAMN15341344</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR12068550</td>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>Transcriptome profiling of Mesoplasma florum s...</td>\n",
       "      <td>Mesoplasma florum is a near-minimal bacterium ...</td>\n",
       "      <td>PRJNA641121</td>\n",
       "      <td>SAMN15341341</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR12954319</td>\n",
       "      <td>Heribaudiella fluviatilis</td>\n",
       "      <td>Sequencing data of the Phaeoexplorer project</td>\n",
       "      <td>This project collects genomic and transcriptom...</td>\n",
       "      <td>PRJEB72149</td>\n",
       "      <td>SAMEA115094190</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id                   organism  \\\n",
       "0  SRR12068547          Mesoplasma florum   \n",
       "1  SRR12068550          Mesoplasma florum   \n",
       "2  ERR12954319  Heribaudiella fluviatilis   \n",
       "\n",
       "                                         study_title  \\\n",
       "0  Transcriptome profiling of Mesoplasma florum s...   \n",
       "1  Transcriptome profiling of Mesoplasma florum s...   \n",
       "2       Sequencing data of the Phaeoexplorer project   \n",
       "\n",
       "                                      study_abstract   bioproject  \\\n",
       "0  Mesoplasma florum is a near-minimal bacterium ...  PRJNA641121   \n",
       "1  Mesoplasma florum is a near-minimal bacterium ...  PRJNA641121   \n",
       "2  This project collects genomic and transcriptom...   PRJEB72149   \n",
       "\n",
       "        biosample library_strategy library_source library_selection  \\\n",
       "0    SAMN15341344             None           None              None   \n",
       "1    SAMN15341341             None           None              None   \n",
       "2  SAMEA115094190             None           None              None   \n",
       "\n",
       "  library_layout instrument run_spots run_bases run_published  \n",
       "0           None   ILLUMINA      None      None          None  \n",
       "1           None   ILLUMINA      None      None          None  \n",
       "2           None   ILLUMINA      None      None          None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Path to your schema SQL file\n",
    "schema_sql_path = Path('../planter/database/schema/migrations/001_initial_schema.sql')\n",
    "\n",
    "# Path for the master (merged) DuckDB\n",
    "master_db_path = '/tmp/merged.duckdb'\n",
    "\n",
    "def merge_duckdbs(\n",
    "    duckdb_paths: List[Union[str, Path]],\n",
    "    master_db_path: Union[str, Path],\n",
    "    schema_sql_path: Union[str, Path]\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    Merge multiple DuckDB databases into a master DuckDB.\n",
    "    \n",
    "    Parameters:\n",
    "      duckdb_paths (List[Union[str, Path]]): List of paths to source DuckDB files.\n",
    "      master_db_path (Union[str, Path]): Path to the master (merged) DuckDB.\n",
    "      schema_sql_path (Union[str, Path]): Path to the SQL file defining the schema.\n",
    "    \n",
    "    The function:\n",
    "      - Creates (or opens) the master database.\n",
    "      - Executes the schema SQL to create tables if they don't exist.\n",
    "      - Iterates through each source database, attaches it,\n",
    "        and inserts data into the master tables in dependency order.\n",
    "      - Uses INSERT OR IGNORE to avoid duplicate key errors.\n",
    "      - Detaches each source database after merging.\n",
    "    \"\"\"\n",
    "    \n",
    "    master_db_path = str(master_db_path)\n",
    "    schema_sql_path = Path(schema_sql_path)\n",
    "    \n",
    "    # Read the schema SQL\n",
    "    schema_sql = schema_sql_path.read_text()\n",
    "    \n",
    "    with duckdb.connect(master_db_path) as master_conn:\n",
    "        # Set up the schema in the master database\n",
    "        master_conn.execute(schema_sql)\n",
    "        \n",
    "        # Process each source DuckDB\n",
    "        for i, source_db in enumerate(duckdb_paths):\n",
    "            alias = f\"db{i}\"\n",
    "            source_db_str = str(source_db)\n",
    "            print(f\"Attaching {source_db_str} as {alias}...\")\n",
    "            master_conn.execute(f\"ATTACH '{source_db_str}' AS {alias};\")\n",
    "            \n",
    "            # Insert data in dependency order\n",
    "            master_conn.execute(f\"\"\"\n",
    "                INSERT OR IGNORE INTO sra_metadata\n",
    "                SELECT * FROM {alias}.sra_metadata;\n",
    "            \"\"\")\n",
    "            master_conn.execute(f\"\"\"\n",
    "                INSERT OR IGNORE INTO sequences\n",
    "                SELECT * FROM {alias}.sequences;\n",
    "            \"\"\")\n",
    "            master_conn.execute(f\"\"\"\n",
    "                INSERT OR IGNORE INTO annotations\n",
    "                SELECT * FROM {alias}.annotations;\n",
    "            \"\"\")\n",
    "            master_conn.execute(f\"\"\"\n",
    "                INSERT OR IGNORE INTO go_terms\n",
    "                SELECT * FROM {alias}.go_terms;\n",
    "            \"\"\")\n",
    "            master_conn.execute(f\"\"\"\n",
    "                INSERT OR IGNORE INTO ec_numbers\n",
    "                SELECT * FROM {alias}.ec_numbers;\n",
    "            \"\"\")\n",
    "            master_conn.execute(f\"\"\"\n",
    "                INSERT OR IGNORE INTO kegg_info\n",
    "                SELECT * FROM {alias}.kegg_info;\n",
    "            \"\"\")\n",
    "            master_conn.execute(f\"\"\"\n",
    "                INSERT OR IGNORE INTO clusters\n",
    "                SELECT * FROM {alias}.clusters;\n",
    "            \"\"\")\n",
    "            master_conn.execute(f\"\"\"\n",
    "                INSERT OR IGNORE INTO cluster_members\n",
    "                SELECT * FROM {alias}.cluster_members;\n",
    "            \"\"\")\n",
    "            \n",
    "            # Optionally, merge schema_version if needed:\n",
    "            # master_conn.execute(f\"\"\"\n",
    "            #     INSERT OR IGNORE INTO schema_version\n",
    "            #     SELECT * FROM {alias}.schema_version;\n",
    "            # \"\"\")\n",
    "            \n",
    "            master_conn.execute(f\"DETACH {alias};\")\n",
    "            print(f\"Finished merging {source_db_str}\\n\")\n",
    "        \n",
    "        # Optional commit; DuckDB auto-commits by default.\n",
    "        master_conn.commit()\n",
    "    \n",
    "    print(\"All databases have been merged into:\", master_db_path)\n",
    "    return master_db_path\n",
    "\n",
    "master_db_path = merge_duckdbs(DUCKDB_PATHS, master_db_path, schema_sql_path)\n",
    "display_tables(master_db_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attaching /mnt/data3/planter_outputs/ERR12954319/ERR12954319.duckdb as db0...\n",
      "Finished merging /mnt/data3/planter_outputs/ERR12954319/ERR12954319.duckdb\n",
      "\n",
      "Attaching /mnt/data3/planter_outputs/SRR12068548/SRR12068548.duckdb as db1...\n",
      "Finished merging /mnt/data3/planter_outputs/SRR12068548/SRR12068548.duckdb\n",
      "\n",
      "Attaching /mnt/data3/planter_outputs/SRR12068552/SRR12068552.duckdb as db2...\n",
      "Finished merging /mnt/data3/planter_outputs/SRR12068552/SRR12068552.duckdb\n",
      "\n",
      "Attaching /mnt/data3/planter_outputs/SRR12068551/SRR12068551.duckdb as db3...\n",
      "Finished merging /mnt/data3/planter_outputs/SRR12068551/SRR12068551.duckdb\n",
      "\n",
      "Attaching /mnt/data3/planter_outputs/SRR12068549/SRR12068549.duckdb as db4...\n",
      "Finished merging /mnt/data3/planter_outputs/SRR12068549/SRR12068549.duckdb\n",
      "\n",
      "Attaching /mnt/data3/planter_outputs/SRR12068550/SRR12068550.duckdb as db5...\n",
      "Finished merging /mnt/data3/planter_outputs/SRR12068550/SRR12068550.duckdb\n",
      "\n",
      "Attaching /mnt/data3/planter_outputs/ERR2040767/ERR2040767.duckdb as db6...\n",
      "Finished merging /mnt/data3/planter_outputs/ERR2040767/ERR2040767.duckdb\n",
      "\n",
      "Attaching /mnt/data3/planter_outputs/SRR12068547/SRR12068547.duckdb as db7...\n",
      "Finished merging /mnt/data3/planter_outputs/SRR12068547/SRR12068547.duckdb\n",
      "\n",
      "Attaching /mnt/data3/planter_outputs/ERR12954325/ERR12954325.duckdb as db8...\n",
      "Finished merging /mnt/data3/planter_outputs/ERR12954325/ERR12954325.duckdb\n",
      "\n",
      "All databases have been merged into: /tmp/merged.duckdb\n",
      "First 3 rows of annotations:\n",
      "annotations has 88743 rows and 10 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggnog_ogs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>cog_category</th>\n",
       "      <th>description</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>265311.Mfl224</td>\n",
       "      <td>1.930000e-06</td>\n",
       "      <td>53.5</td>\n",
       "      <td>COG1564@1|root,COG1564@2|Bacteria,3WTAN@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>H</td>\n",
       "      <td>Thiamin pyrophosphokinase, vitamin B1 binding ...</td>\n",
       "      <td>thiN</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>265311.Mfl040</td>\n",
       "      <td>3.740000e-45</td>\n",
       "      <td>155.0</td>\n",
       "      <td>COG0022@1|root,COG0022@2|Bacteria,3WSYT@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>C</td>\n",
       "      <td>dehydrogenase e1 component</td>\n",
       "      <td>pdhB</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...</td>\n",
       "      <td>265311.Mfl489</td>\n",
       "      <td>1.300000e-29</td>\n",
       "      <td>120.0</td>\n",
       "      <td>2EWBJ@1|root,33PQA@2|Bacteria,3WT4V@544448|Ten...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  seed_ortholog  \\\n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...  265311.Mfl224   \n",
       "1  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...  265311.Mfl040   \n",
       "2  v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...  265311.Mfl489   \n",
       "\n",
       "         evalue  score                                         eggnog_ogs  \\\n",
       "0  1.930000e-06   53.5  COG1564@1|root,COG1564@2|Bacteria,3WTAN@544448...   \n",
       "1  3.740000e-45  155.0  COG0022@1|root,COG0022@2|Bacteria,3WSYT@544448...   \n",
       "2  1.300000e-29  120.0  2EWBJ@1|root,33PQA@2|Bacteria,3WT4V@544448|Ten...   \n",
       "\n",
       "        max_annot_lvl cog_category  \\\n",
       "0  544448|Tenericutes            H   \n",
       "1  544448|Tenericutes            C   \n",
       "2  544448|Tenericutes            -   \n",
       "\n",
       "                                         description preferred_name  \\\n",
       "0  Thiamin pyrophosphokinase, vitamin B1 binding ...           thiN   \n",
       "1                         dehydrogenase e1 component           pdhB   \n",
       "2                                                  -              -   \n",
       "\n",
       "     sample_id  \n",
       "0  SRR12068547  \n",
       "1  SRR12068547  \n",
       "2  SRR12068547  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of cluster_members:\n",
      "cluster_members has 0 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>cluster_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, cluster_id]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of clusters:\n",
      "clusters has 0 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>representative_seqhash_id</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cluster_id, representative_seqhash_id, size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of ec_numbers:\n",
      "ec_numbers has 31634 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_05fc6dc2fcc79581a85e20a59498e2ca4389b43...</td>\n",
       "      <td>1.2.4.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_116aa24751cc0c03d972d124b07632ab5a2db0e...</td>\n",
       "      <td>3.4.23.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_166a64b63f322440c5a41bc6219f0a942373c2b...</td>\n",
       "      <td>5.4.2.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  ec_number\n",
       "0  v1_DLS_05fc6dc2fcc79581a85e20a59498e2ca4389b43...    1.2.4.4\n",
       "1  v1_DLS_116aa24751cc0c03d972d124b07632ab5a2db0e...  3.4.23.36\n",
       "2  v1_DLS_166a64b63f322440c5a41bc6219f0a942373c2b...   5.4.2.12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of go_terms:\n",
      "go_terms has 3756450 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>go_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>GO:0140032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_1016173829ac36ac2333823a671a085347bddb2...</td>\n",
       "      <td>GO:0005515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_27169752765a0e3a22d9727f0dcb6ece7c520a9...</td>\n",
       "      <td>GO:0005622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id     go_term\n",
       "0  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...  GO:0140032\n",
       "1  v1_DLS_1016173829ac36ac2333823a671a085347bddb2...  GO:0005515\n",
       "2  v1_DLS_27169752765a0e3a22d9727f0dcb6ece7c520a9...  GO:0005622"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of kegg_info:\n",
      "kegg_info has 0 rows and 6 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>kegg_ko</th>\n",
       "      <th>kegg_pathway</th>\n",
       "      <th>kegg_module</th>\n",
       "      <th>kegg_reaction</th>\n",
       "      <th>kegg_rclass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, kegg_ko, kegg_pathway, kegg_module, kegg_reaction, kegg_rclass]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of schema_version:\n",
      "schema_version has 0 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>version</th>\n",
       "      <th>migration_name</th>\n",
       "      <th>applied_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [version, migration_name, applied_at]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sequences:\n",
      "sequences has 149833 rows and 7 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>assembly_date</th>\n",
       "      <th>is_representative</th>\n",
       "      <th>repseq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774568</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...</td>\n",
       "      <td>MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774589</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774599</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  \\\n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...   \n",
       "1  v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...   \n",
       "2  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...   \n",
       "\n",
       "                                            sequence    sample_id  \\\n",
       "0  MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  SRR12068547   \n",
       "1  MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  SRR12068547   \n",
       "2  MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...  SRR12068547   \n",
       "\n",
       "               assembly_date  is_representative  \\\n",
       "0 2025-02-11 02:04:12.774568              False   \n",
       "1 2025-02-11 02:04:12.774589              False   \n",
       "2 2025-02-11 02:04:12.774599              False   \n",
       "\n",
       "                                           repseq_id  length  \n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...     160  \n",
       "1  v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...     229  \n",
       "2  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...     117  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sra_metadata:\n",
      "sra_metadata has 9 rows and 14 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_abstract</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>instrument</th>\n",
       "      <th>run_spots</th>\n",
       "      <th>run_bases</th>\n",
       "      <th>run_published</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>Transcriptome profiling of Mesoplasma florum s...</td>\n",
       "      <td>Mesoplasma florum is a near-minimal bacterium ...</td>\n",
       "      <td>PRJNA641121</td>\n",
       "      <td>SAMN15341344</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR12068550</td>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>Transcriptome profiling of Mesoplasma florum s...</td>\n",
       "      <td>Mesoplasma florum is a near-minimal bacterium ...</td>\n",
       "      <td>PRJNA641121</td>\n",
       "      <td>SAMN15341341</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR12954319</td>\n",
       "      <td>Heribaudiella fluviatilis</td>\n",
       "      <td>Sequencing data of the Phaeoexplorer project</td>\n",
       "      <td>This project collects genomic and transcriptom...</td>\n",
       "      <td>PRJEB72149</td>\n",
       "      <td>SAMEA115094190</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id                   organism  \\\n",
       "0  SRR12068547          Mesoplasma florum   \n",
       "1  SRR12068550          Mesoplasma florum   \n",
       "2  ERR12954319  Heribaudiella fluviatilis   \n",
       "\n",
       "                                         study_title  \\\n",
       "0  Transcriptome profiling of Mesoplasma florum s...   \n",
       "1  Transcriptome profiling of Mesoplasma florum s...   \n",
       "2       Sequencing data of the Phaeoexplorer project   \n",
       "\n",
       "                                      study_abstract   bioproject  \\\n",
       "0  Mesoplasma florum is a near-minimal bacterium ...  PRJNA641121   \n",
       "1  Mesoplasma florum is a near-minimal bacterium ...  PRJNA641121   \n",
       "2  This project collects genomic and transcriptom...   PRJEB72149   \n",
       "\n",
       "        biosample library_strategy library_source library_selection  \\\n",
       "0    SAMN15341344             None           None              None   \n",
       "1    SAMN15341341             None           None              None   \n",
       "2  SAMEA115094190             None           None              None   \n",
       "\n",
       "  library_layout instrument run_spots run_bases run_published  \n",
       "0           None   ILLUMINA      None      None          None  \n",
       "1           None   ILLUMINA      None      None          None  \n",
       "2           None   ILLUMINA      None      None          None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "duckdb_paths = list(Path('/mnt/data3/planter_outputs').rglob('*RR*.duckdb'))\n",
    "master_db_path = merge_duckdbs(duckdb_paths, master_db_path, schema_sql_path)\n",
    "display_tables(master_db_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of annotations:\n",
      "annotations has 88743 rows and 10 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggnog_ogs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>cog_category</th>\n",
       "      <th>description</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>265311.Mfl224</td>\n",
       "      <td>1.930000e-06</td>\n",
       "      <td>53.5</td>\n",
       "      <td>COG1564@1|root,COG1564@2|Bacteria,3WTAN@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>H</td>\n",
       "      <td>Thiamin pyrophosphokinase, vitamin B1 binding ...</td>\n",
       "      <td>thiN</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>265311.Mfl040</td>\n",
       "      <td>3.740000e-45</td>\n",
       "      <td>155.0</td>\n",
       "      <td>COG0022@1|root,COG0022@2|Bacteria,3WSYT@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>C</td>\n",
       "      <td>dehydrogenase e1 component</td>\n",
       "      <td>pdhB</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...</td>\n",
       "      <td>265311.Mfl489</td>\n",
       "      <td>1.300000e-29</td>\n",
       "      <td>120.0</td>\n",
       "      <td>2EWBJ@1|root,33PQA@2|Bacteria,3WT4V@544448|Ten...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  seed_ortholog  \\\n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...  265311.Mfl224   \n",
       "1  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...  265311.Mfl040   \n",
       "2  v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...  265311.Mfl489   \n",
       "\n",
       "         evalue  score                                         eggnog_ogs  \\\n",
       "0  1.930000e-06   53.5  COG1564@1|root,COG1564@2|Bacteria,3WTAN@544448...   \n",
       "1  3.740000e-45  155.0  COG0022@1|root,COG0022@2|Bacteria,3WSYT@544448...   \n",
       "2  1.300000e-29  120.0  2EWBJ@1|root,33PQA@2|Bacteria,3WT4V@544448|Ten...   \n",
       "\n",
       "        max_annot_lvl cog_category  \\\n",
       "0  544448|Tenericutes            H   \n",
       "1  544448|Tenericutes            C   \n",
       "2  544448|Tenericutes            -   \n",
       "\n",
       "                                         description preferred_name  \\\n",
       "0  Thiamin pyrophosphokinase, vitamin B1 binding ...           thiN   \n",
       "1                         dehydrogenase e1 component           pdhB   \n",
       "2                                                  -              -   \n",
       "\n",
       "     sample_id  \n",
       "0  SRR12068547  \n",
       "1  SRR12068547  \n",
       "2  SRR12068547  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of cluster_members:\n",
      "cluster_members has 0 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>cluster_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, cluster_id]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of clusters:\n",
      "clusters has 0 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>representative_seqhash_id</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cluster_id, representative_seqhash_id, size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of ec_numbers:\n",
      "ec_numbers has 31634 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_05fc6dc2fcc79581a85e20a59498e2ca4389b43...</td>\n",
       "      <td>1.2.4.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_116aa24751cc0c03d972d124b07632ab5a2db0e...</td>\n",
       "      <td>3.4.23.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_166a64b63f322440c5a41bc6219f0a942373c2b...</td>\n",
       "      <td>5.4.2.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  ec_number\n",
       "0  v1_DLS_05fc6dc2fcc79581a85e20a59498e2ca4389b43...    1.2.4.4\n",
       "1  v1_DLS_116aa24751cc0c03d972d124b07632ab5a2db0e...  3.4.23.36\n",
       "2  v1_DLS_166a64b63f322440c5a41bc6219f0a942373c2b...   5.4.2.12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of go_terms:\n",
      "go_terms has 3756450 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>go_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>GO:0140032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_1016173829ac36ac2333823a671a085347bddb2...</td>\n",
       "      <td>GO:0005515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_27169752765a0e3a22d9727f0dcb6ece7c520a9...</td>\n",
       "      <td>GO:0005622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id     go_term\n",
       "0  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...  GO:0140032\n",
       "1  v1_DLS_1016173829ac36ac2333823a671a085347bddb2...  GO:0005515\n",
       "2  v1_DLS_27169752765a0e3a22d9727f0dcb6ece7c520a9...  GO:0005622"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of kegg_info:\n",
      "kegg_info has 0 rows and 6 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>kegg_ko</th>\n",
       "      <th>kegg_pathway</th>\n",
       "      <th>kegg_module</th>\n",
       "      <th>kegg_reaction</th>\n",
       "      <th>kegg_rclass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, kegg_ko, kegg_pathway, kegg_module, kegg_reaction, kegg_rclass]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of schema_version:\n",
      "schema_version has 0 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>version</th>\n",
       "      <th>migration_name</th>\n",
       "      <th>applied_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [version, migration_name, applied_at]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sequences:\n",
      "sequences has 149833 rows and 7 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>assembly_date</th>\n",
       "      <th>is_representative</th>\n",
       "      <th>repseq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774568</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...</td>\n",
       "      <td>MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774589</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774599</td>\n",
       "      <td>False</td>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  \\\n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...   \n",
       "1  v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...   \n",
       "2  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...   \n",
       "\n",
       "                                            sequence    sample_id  \\\n",
       "0  MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  SRR12068547   \n",
       "1  MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  SRR12068547   \n",
       "2  MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...  SRR12068547   \n",
       "\n",
       "               assembly_date  is_representative  \\\n",
       "0 2025-02-11 02:04:12.774568              False   \n",
       "1 2025-02-11 02:04:12.774589              False   \n",
       "2 2025-02-11 02:04:12.774599              False   \n",
       "\n",
       "                                           repseq_id  length  \n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...     160  \n",
       "1  v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...     229  \n",
       "2  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...     117  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sra_metadata:\n",
      "sra_metadata has 9 rows and 14 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_abstract</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>instrument</th>\n",
       "      <th>run_spots</th>\n",
       "      <th>run_bases</th>\n",
       "      <th>run_published</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>Transcriptome profiling of Mesoplasma florum s...</td>\n",
       "      <td>Mesoplasma florum is a near-minimal bacterium ...</td>\n",
       "      <td>PRJNA641121</td>\n",
       "      <td>SAMN15341344</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR12068550</td>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>Transcriptome profiling of Mesoplasma florum s...</td>\n",
       "      <td>Mesoplasma florum is a near-minimal bacterium ...</td>\n",
       "      <td>PRJNA641121</td>\n",
       "      <td>SAMN15341341</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR12954319</td>\n",
       "      <td>Heribaudiella fluviatilis</td>\n",
       "      <td>Sequencing data of the Phaeoexplorer project</td>\n",
       "      <td>This project collects genomic and transcriptom...</td>\n",
       "      <td>PRJEB72149</td>\n",
       "      <td>SAMEA115094190</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id                   organism  \\\n",
       "0  SRR12068547          Mesoplasma florum   \n",
       "1  SRR12068550          Mesoplasma florum   \n",
       "2  ERR12954319  Heribaudiella fluviatilis   \n",
       "\n",
       "                                         study_title  \\\n",
       "0  Transcriptome profiling of Mesoplasma florum s...   \n",
       "1  Transcriptome profiling of Mesoplasma florum s...   \n",
       "2       Sequencing data of the Phaeoexplorer project   \n",
       "\n",
       "                                      study_abstract   bioproject  \\\n",
       "0  Mesoplasma florum is a near-minimal bacterium ...  PRJNA641121   \n",
       "1  Mesoplasma florum is a near-minimal bacterium ...  PRJNA641121   \n",
       "2  This project collects genomic and transcriptom...   PRJEB72149   \n",
       "\n",
       "        biosample library_strategy library_source library_selection  \\\n",
       "0    SAMN15341344             None           None              None   \n",
       "1    SAMN15341341             None           None              None   \n",
       "2  SAMEA115094190             None           None              None   \n",
       "\n",
       "  library_layout instrument run_spots run_bases run_published  \n",
       "0           None   ILLUMINA      None      None          None  \n",
       "1           None   ILLUMINA      None      None          None  \n",
       "2           None   ILLUMINA      None      None          None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display_tables(master_db_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repseq_id</th>\n",
       "      <th>seqhash_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_693b34809e671efcc491066951624176abdec2e...</td>\n",
       "      <td>v1_DLS_693b34809e671efcc491066951624176abdec2e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_693b34809e671efcc491066951624176abdec2e...</td>\n",
       "      <td>v1_DLS_693b34809e671efcc491066951624176abdec2e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_693b34809e671efcc491066951624176abdec2e...</td>\n",
       "      <td>v1_DLS_754e0acb8250a3a8f1f6f8ca380d0fb660f7008...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>v1_DLS_693b34809e671efcc491066951624176abdec2e...</td>\n",
       "      <td>v1_DLS_d1e220b01aa8d311108099006dccc45340eddd1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>v1_DLS_08d72bfa8cd748621c47c5ab9976e310ad1393f...</td>\n",
       "      <td>v1_DLS_08d72bfa8cd748621c47c5ab9976e310ad1393f...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           repseq_id  \\\n",
       "0  v1_DLS_693b34809e671efcc491066951624176abdec2e...   \n",
       "1  v1_DLS_693b34809e671efcc491066951624176abdec2e...   \n",
       "2  v1_DLS_693b34809e671efcc491066951624176abdec2e...   \n",
       "3  v1_DLS_693b34809e671efcc491066951624176abdec2e...   \n",
       "4  v1_DLS_08d72bfa8cd748621c47c5ab9976e310ad1393f...   \n",
       "\n",
       "                                          seqhash_id  \n",
       "0  v1_DLS_693b34809e671efcc491066951624176abdec2e...  \n",
       "1  v1_DLS_693b34809e671efcc491066951624176abdec2e...  \n",
       "2  v1_DLS_754e0acb8250a3a8f1f6f8ca380d0fb660f7008...  \n",
       "3  v1_DLS_d1e220b01aa8d311108099006dccc45340eddd1...  \n",
       "4  v1_DLS_08d72bfa8cd748621c47c5ab9976e310ad1393f...  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(73065, 2)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "56341"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repseq_clusters = pd.read_csv('/mnt/data3/planter_outputs/cluster/newClusterDB.tsv', delimiter='\\t', header=None, names=['repseq_id', 'seqhash_id'])\n",
    "repseq_clusters.head()\n",
    "repseq_clusters.shape\n",
    "repseq_clusters['repseq_id'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Master DB updated using cluster mapping.\n"
     ]
    }
   ],
   "source": [
    "def update_master_db_with_cluster_mapping(master_db_path: str, tsv_path: str, inplace=False):\n",
    "    \"\"\"\n",
    "    Update the 'sequences' table in the master DuckDB using a cluster mapping TSV.\n",
    "    \n",
    "    Parameters:\n",
    "        master_db_path (str): Path to the master DuckDB file.\n",
    "        tsv_path (str): Path to the TSV file containing cluster mappings.\n",
    "                        The TSV should have two columns (no header):\n",
    "                          - Column 0: repseq_id (the representative seqhash)\n",
    "                          - Column 1: seqhash_id (the member seqhash)\n",
    "                          \n",
    "    The function updates the 'repseq_id' column in the 'sequences' table such that\n",
    "    for every row where sequences.seqhash_id matches the TSV's member, repseq_id is set\n",
    "    to the TSV's repseq_id. It also sets is_representative to TRUE for sequences that are representatives.\n",
    "    \"\"\"\n",
    "    # Connect to the master DuckDB.\n",
    "    con = duckdb.connect(master_db_path)\n",
    "    \n",
    "    # Create a temporary table by reading the TSV file.\n",
    "    # We assume the TSV has no header and uses tab as the separator.\n",
    "    con.execute(f\"\"\"\n",
    "    CREATE TEMPORARY TABLE new_clusters AS \n",
    "      SELECT * FROM read_csv_auto('{tsv_path}', header=False, sep='\\t', names=['repseq_id', 'seqhash_id'])\n",
    "      AS (repseq_id VARCHAR, seqhash_id VARCHAR);\n",
    "    \"\"\")\n",
    "    \n",
    "    # Update the sequences table: set repseq_id from new_clusters for matching seqhash_id.\n",
    "    # con.execute(\"\"\"\n",
    "    # UPDATE sequences\n",
    "    # SET repseq_id = new_clusters.repseq_id\n",
    "    # FROM new_clusters\n",
    "    # WHERE sequences.seqhash_id = new_clusters.seqhash_id;\n",
    "    # \"\"\")\n",
    "    \n",
    "    # # Optionally, update the is_representative flag.\n",
    "    # # For every sequence that appears as a repseq_id in the new_clusters table, mark it as representative.\n",
    "    con.execute(\"\"\"\n",
    "    UPDATE sequences\n",
    "    SET is_representative = TRUE\n",
    "    WHERE seqhash_id IN (\n",
    "        SELECT DISTINCT repseq_id FROM new_clusters\n",
    "    );\n",
    "    \"\"\")\n",
    "    \n",
    "    con.close()\n",
    "    print(\"Master DB updated using cluster mapping.\")\n",
    "\n",
    "update_master_db_with_cluster_mapping(master_db_path, '/mnt/data3/planter_outputs/cluster/newClusterDB.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of annotations:\n",
      "annotations has 88743 rows and 10 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggnog_ogs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>cog_category</th>\n",
       "      <th>description</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>265311.Mfl224</td>\n",
       "      <td>1.930000e-06</td>\n",
       "      <td>53.5</td>\n",
       "      <td>COG1564@1|root,COG1564@2|Bacteria,3WTAN@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>H</td>\n",
       "      <td>Thiamin pyrophosphokinase, vitamin B1 binding ...</td>\n",
       "      <td>thiN</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>265311.Mfl040</td>\n",
       "      <td>3.740000e-45</td>\n",
       "      <td>155.0</td>\n",
       "      <td>COG0022@1|root,COG0022@2|Bacteria,3WSYT@544448...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>C</td>\n",
       "      <td>dehydrogenase e1 component</td>\n",
       "      <td>pdhB</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...</td>\n",
       "      <td>265311.Mfl489</td>\n",
       "      <td>1.300000e-29</td>\n",
       "      <td>120.0</td>\n",
       "      <td>2EWBJ@1|root,33PQA@2|Bacteria,3WT4V@544448|Ten...</td>\n",
       "      <td>544448|Tenericutes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR12068547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  seed_ortholog  \\\n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...  265311.Mfl224   \n",
       "1  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...  265311.Mfl040   \n",
       "2  v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...  265311.Mfl489   \n",
       "\n",
       "         evalue  score                                         eggnog_ogs  \\\n",
       "0  1.930000e-06   53.5  COG1564@1|root,COG1564@2|Bacteria,3WTAN@544448...   \n",
       "1  3.740000e-45  155.0  COG0022@1|root,COG0022@2|Bacteria,3WSYT@544448...   \n",
       "2  1.300000e-29  120.0  2EWBJ@1|root,33PQA@2|Bacteria,3WT4V@544448|Ten...   \n",
       "\n",
       "        max_annot_lvl cog_category  \\\n",
       "0  544448|Tenericutes            H   \n",
       "1  544448|Tenericutes            C   \n",
       "2  544448|Tenericutes            -   \n",
       "\n",
       "                                         description preferred_name  \\\n",
       "0  Thiamin pyrophosphokinase, vitamin B1 binding ...           thiN   \n",
       "1                         dehydrogenase e1 component           pdhB   \n",
       "2                                                  -              -   \n",
       "\n",
       "     sample_id  \n",
       "0  SRR12068547  \n",
       "1  SRR12068547  \n",
       "2  SRR12068547  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of cluster_members:\n",
      "cluster_members has 0 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>cluster_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, cluster_id]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of clusters:\n",
      "clusters has 0 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>representative_seqhash_id</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cluster_id, representative_seqhash_id, size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of ec_numbers:\n",
      "ec_numbers has 31634 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_05fc6dc2fcc79581a85e20a59498e2ca4389b43...</td>\n",
       "      <td>1.2.4.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_116aa24751cc0c03d972d124b07632ab5a2db0e...</td>\n",
       "      <td>3.4.23.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_166a64b63f322440c5a41bc6219f0a942373c2b...</td>\n",
       "      <td>5.4.2.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  ec_number\n",
       "0  v1_DLS_05fc6dc2fcc79581a85e20a59498e2ca4389b43...    1.2.4.4\n",
       "1  v1_DLS_116aa24751cc0c03d972d124b07632ab5a2db0e...  3.4.23.36\n",
       "2  v1_DLS_166a64b63f322440c5a41bc6219f0a942373c2b...   5.4.2.12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of go_terms:\n",
      "go_terms has 3756450 rows and 2 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>go_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>GO:0140032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_1016173829ac36ac2333823a671a085347bddb2...</td>\n",
       "      <td>GO:0005515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_27169752765a0e3a22d9727f0dcb6ece7c520a9...</td>\n",
       "      <td>GO:0005622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id     go_term\n",
       "0  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...  GO:0140032\n",
       "1  v1_DLS_1016173829ac36ac2333823a671a085347bddb2...  GO:0005515\n",
       "2  v1_DLS_27169752765a0e3a22d9727f0dcb6ece7c520a9...  GO:0005622"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of kegg_info:\n",
      "kegg_info has 0 rows and 6 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>kegg_ko</th>\n",
       "      <th>kegg_pathway</th>\n",
       "      <th>kegg_module</th>\n",
       "      <th>kegg_reaction</th>\n",
       "      <th>kegg_rclass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqhash_id, kegg_ko, kegg_pathway, kegg_module, kegg_reaction, kegg_rclass]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of schema_version:\n",
      "schema_version has 0 rows and 3 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>version</th>\n",
       "      <th>migration_name</th>\n",
       "      <th>applied_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [version, migration_name, applied_at]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sequences:\n",
      "sequences has 149833 rows and 7 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>assembly_date</th>\n",
       "      <th>is_representative</th>\n",
       "      <th>repseq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774568</td>\n",
       "      <td>True</td>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...</td>\n",
       "      <td>MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774589</td>\n",
       "      <td>True</td>\n",
       "      <td>v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...</td>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>2025-02-11 02:04:12.774599</td>\n",
       "      <td>True</td>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          seqhash_id  \\\n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...   \n",
       "1  v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...   \n",
       "2  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...   \n",
       "\n",
       "                                            sequence    sample_id  \\\n",
       "0  MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  SRR12068547   \n",
       "1  MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  SRR12068547   \n",
       "2  MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...  SRR12068547   \n",
       "\n",
       "               assembly_date  is_representative  \\\n",
       "0 2025-02-11 02:04:12.774568               True   \n",
       "1 2025-02-11 02:04:12.774589               True   \n",
       "2 2025-02-11 02:04:12.774599               True   \n",
       "\n",
       "                                           repseq_id  length  \n",
       "0  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...     160  \n",
       "1  v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...     229  \n",
       "2  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...     117  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 3 rows of sra_metadata:\n",
      "sra_metadata has 9 rows and 14 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_abstract</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>instrument</th>\n",
       "      <th>run_spots</th>\n",
       "      <th>run_bases</th>\n",
       "      <th>run_published</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR12068547</td>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>Transcriptome profiling of Mesoplasma florum s...</td>\n",
       "      <td>Mesoplasma florum is a near-minimal bacterium ...</td>\n",
       "      <td>PRJNA641121</td>\n",
       "      <td>SAMN15341344</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR12068550</td>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>Transcriptome profiling of Mesoplasma florum s...</td>\n",
       "      <td>Mesoplasma florum is a near-minimal bacterium ...</td>\n",
       "      <td>PRJNA641121</td>\n",
       "      <td>SAMN15341341</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR12954319</td>\n",
       "      <td>Heribaudiella fluviatilis</td>\n",
       "      <td>Sequencing data of the Phaeoexplorer project</td>\n",
       "      <td>This project collects genomic and transcriptom...</td>\n",
       "      <td>PRJEB72149</td>\n",
       "      <td>SAMEA115094190</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id                   organism  \\\n",
       "0  SRR12068547          Mesoplasma florum   \n",
       "1  SRR12068550          Mesoplasma florum   \n",
       "2  ERR12954319  Heribaudiella fluviatilis   \n",
       "\n",
       "                                         study_title  \\\n",
       "0  Transcriptome profiling of Mesoplasma florum s...   \n",
       "1  Transcriptome profiling of Mesoplasma florum s...   \n",
       "2       Sequencing data of the Phaeoexplorer project   \n",
       "\n",
       "                                      study_abstract   bioproject  \\\n",
       "0  Mesoplasma florum is a near-minimal bacterium ...  PRJNA641121   \n",
       "1  Mesoplasma florum is a near-minimal bacterium ...  PRJNA641121   \n",
       "2  This project collects genomic and transcriptom...   PRJEB72149   \n",
       "\n",
       "        biosample library_strategy library_source library_selection  \\\n",
       "0    SAMN15341344             None           None              None   \n",
       "1    SAMN15341341             None           None              None   \n",
       "2  SAMEA115094190             None           None              None   \n",
       "\n",
       "  library_layout instrument run_spots run_bases run_published  \n",
       "0           None   ILLUMINA      None      None          None  \n",
       "1           None   ILLUMINA      None      None          None  \n",
       "2           None   ILLUMINA      None      None          None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display_tables(master_db_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "with duckdb.connect(master_db_path) as con:\n",
    "    df = con.execute(\"\"\"\n",
    "    SELECT m.organism, s.seqhash_id, s.sequence\n",
    "    FROM sequences s\n",
    "    JOIN sra_metadata m ON s.sample_id = m.sample_id\n",
    "    WHERE s.is_representative = TRUE;\n",
    "    \"\"\").fetchdf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>seqhash_id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>v1_DLS_0076d08a43a612188f215c23180111039b05c1b...</td>\n",
       "      <td>MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...</td>\n",
       "      <td>MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...</td>\n",
       "      <td>MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...</td>\n",
       "      <td>MYDFIPITVPNNNXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mesoplasma florum</td>\n",
       "      <td>v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...</td>\n",
       "      <td>MFYTTTKKQFHLLKDYKLNWQXXXXXXXXXXXXXXXXXXXXXXXXX...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56336</th>\n",
       "      <td>Ascophyllum nodosum</td>\n",
       "      <td>v1_DLS_708463fc43ff8b43c03f893d2a71dba98a21425...</td>\n",
       "      <td>MFSSVFDKVVSATAPDALATFPPRAALPVPPSTLCCLSFVDRFLTH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56337</th>\n",
       "      <td>Ascophyllum nodosum</td>\n",
       "      <td>v1_DLS_70854bbb50bcc05f9a40dcbb148cc5e4c903011...</td>\n",
       "      <td>MRKPILVPYWSTWNYSIRIVSDQRSEGVCVVLGSPPFAVRGTSSLS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56338</th>\n",
       "      <td>Ascophyllum nodosum</td>\n",
       "      <td>v1_DLS_708728a6e86ec522e7e23e2e24113ad92934efa...</td>\n",
       "      <td>MSGLTRGETAEPVSRDQILRLEGIQERRAAARQANRDAAMDRKERE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56339</th>\n",
       "      <td>Ascophyllum nodosum</td>\n",
       "      <td>v1_DLS_708728a6e86ec522e7e23e2e24113ad92934efa...</td>\n",
       "      <td>MTAGRRVKPATMIDRAIAVVRDFNPATQTVDSYADDALLGGNSKTS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56340</th>\n",
       "      <td>Ascophyllum nodosum</td>\n",
       "      <td>v1_DLS_708b746f66d551f1bd727a48e7567fae5e5fa76...</td>\n",
       "      <td>MSQEYQQGGLASLKHGADMGSRASCTVIDPELAIARQSFPDESLAW...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56341 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  organism                                         seqhash_id  \\\n",
       "0        Mesoplasma florum  v1_DLS_0076d08a43a612188f215c23180111039b05c1b...   \n",
       "1        Mesoplasma florum  v1_DLS_00dba5ffc2660fb5748dc31acb772b2ad64d737...   \n",
       "2        Mesoplasma florum  v1_DLS_01dd7c83e5a125c8fdd52f600ed3662e13d57e7...   \n",
       "3        Mesoplasma florum  v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...   \n",
       "4        Mesoplasma florum  v1_DLS_02340e54c7bf3192db0a8f6d8bad893802d33fb...   \n",
       "...                    ...                                                ...   \n",
       "56336  Ascophyllum nodosum  v1_DLS_708463fc43ff8b43c03f893d2a71dba98a21425...   \n",
       "56337  Ascophyllum nodosum  v1_DLS_70854bbb50bcc05f9a40dcbb148cc5e4c903011...   \n",
       "56338  Ascophyllum nodosum  v1_DLS_708728a6e86ec522e7e23e2e24113ad92934efa...   \n",
       "56339  Ascophyllum nodosum  v1_DLS_708728a6e86ec522e7e23e2e24113ad92934efa...   \n",
       "56340  Ascophyllum nodosum  v1_DLS_708b746f66d551f1bd727a48e7567fae5e5fa76...   \n",
       "\n",
       "                                                sequence  \n",
       "0      MXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  \n",
       "1      MKPKWFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  \n",
       "2      MVIRMPMGGGIRALEHHSEALEAIFAHIPGIKTVMPSTPYDTKGLL...  \n",
       "3      MYDFIPITVPNNNXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...  \n",
       "4      MFYTTTKKQFHLLKDYKLNWQXXXXXXXXXXXXXXXXXXXXXXXXX...  \n",
       "...                                                  ...  \n",
       "56336  MFSSVFDKVVSATAPDALATFPPRAALPVPPSTLCCLSFVDRFLTH...  \n",
       "56337  MRKPILVPYWSTWNYSIRIVSDQRSEGVCVVLGSPPFAVRGTSSLS...  \n",
       "56338  MSGLTRGETAEPVSRDQILRLEGIQERRAAARQANRDAAMDRKERE...  \n",
       "56339  MTAGRRVKPATMIDRAIAVVRDFNPATQTVDSYADDALLGGNSKTS...  \n",
       "56340  MSQEYQQGGLASLKHGADMGSRASCTVIDPELAIARQSFPDESLAW...  \n",
       "\n",
       "[56341 rows x 3 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
